pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,410 @@
+// as-is to the specifed directory or a temporary directory.  The cache directory's contents
+// can be removed by calling the returned BlobCache()'s ClearCache() method.  If a temporary
+// directory is being used, ClearCache() will remove it as well.
+func NewBlobCache(ref types.ImageReference, directories []string) (BlobCache, error) {
+	removeOnCleanup := false
+	if len(directories) == 0 || directories[0] == "" {
+		directories = []string{""}
+		removeOnCleanup = true
+	}
+	return &blobCacheReference{
+		transport:                     Transport,
+		reference:                     ref,
+		directories:                   directories,
+		removeFirstDirectoryOnCleanup: removeOnCleanup,
+	}, nil
+}
+
+func (r *blobCacheReference) Transport() types.ImageTransport {
+	return r.transport
+}
+
+func (r *blobCacheReference) StringWithinTransport() string {
+	directories := strings.Join(r.directories, ",")
+	if directories != "" {
+		return directories + ":" + transports.ImageName(r.reference)
+	}
+	return transports.ImageName(r.reference)
+}
+
+func (r *blobCacheReference) DockerReference() reference.Named {
+	return r.reference.DockerReference()
+}
+
+func (r *blobCacheReference) PolicyConfigurationIdentity() string {
+	return r.reference.PolicyConfigurationIdentity()
+}
+
+func (r *blobCacheReference) PolicyConfigurationNamespaces() []string {
+	return r.reference.PolicyConfigurationNamespaces()
+}
+
+func (r *blobCacheReference) DeleteImage(ctx context.Context, sys *types.SystemContext) error {
+	return r.reference.DeleteImage(ctx, sys)
+}
+
+func (r *blobCacheReference) HasBlob(blobinfo types.BlobInfo) (bool, int64, error) {
+	if blobinfo.Digest == "" {
+		return false, -1, nil
+	}
+
+	for _, directory := range r.directories {
+		if directory == "" {
+			continue
+		}
+		for _, isConfig := range []bool{false, true} {
+			filename := filepath.Join(directory, makeFilename(blobinfo.Digest, isConfig))
+			fileInfo, err := os.Stat(filename)
+			if err == nil && (blobinfo.Size == -1 || blobinfo.Size == fileInfo.Size()) {
+				return true, fileInfo.Size(), nil
+			}
+			if !os.IsNotExist(err) {
+				return false, -1, errors.Wrapf(err, "error checking size of %q", filename)
+			}
+		}
+	}
+
+	return false, -1, nil
+}
+
+func (r *blobCacheReference) Directories() []string {
+	return append([]string{}, r.directories...)
+}
+
+func (r *blobCacheReference) ClearCache() error {
+	for i, directory := range r.directories {
+		// We "own" the directory.
+		if i == 0 && r.removeFirstDirectoryOnCleanup && directory != "" {
+			if err := os.RemoveAll(directory); err != nil {
+				return errors.Wrapf(err, "error removing %q", directory)
+			}
+			r.directories[0] = ""
+			continue
+		}
+		// We don't "own" this directory, so just clear its contents.
+		f, err := os.Open(directory)
+		if err != nil {
+			return errors.Wrapf(err, "error opening directory %q", directory)
+		}
+		names, err := f.Readdirnames(-1)
+		if err != nil {
+			f.Close()
```

Comment:
Can't we use defer f.Close

Reply:
If we're opening _n_ directories, we'll hold those descriptors open until we've finished cleaning them all.  Probably not worth worrying about that, though, so changing it.